cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -988,70 +988,28 @@ func getNotRegisteredNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances ma
+func (csr *ClusterStateRegistry) getCloudProviderDeletedNodes(allNodes []*apiv1.Node) []*apiv1.Node {
 	nodesRemoved := make([]*apiv1.Node, 0)
-	currentCloudInstances := make(map[string]string, 0)
-	registeredNodes := make(map[string]*apiv1.Node, 0)
-	if len(allNodes) > 0 {
-		_, err := csr.cloudProvider.NodeExists(allNodes[0])
-		// Check if the cloud provider implements nodeExists method
-		nodeExistsNotImplemented := errors.Is(err, cloudprovider.ErrNotImplemented)
-		if nodeExistsNotImplemented {
-			// Fall-back to taint-based node deletion
-			for _, node := range allNodes {
-				if deletetaint.HasToBeDeletedTaint(node) {
-					nodesRemoved = append(nodesRemoved, node)
-				}
-			}
-		} else {
-			for nodeGroupName, instances := range cloudProviderNodeInstances {
-				for _, instance := range instances {
-					currentCloudInstances[instance.Id] = nodeGroupName
-				}
-			}
-			for _, node := range allNodes {
-				registeredNodes[node.Name] = node
-			}
-
-			// Fill previously deleted nodes, if they are still registered in Kubernetes
-			for nodeName, node := range csr.deletedNodes {
-				// Safety check to prevent flagging Kubernetes nodes as deleted
-				// if the Cloud Provider instance is re-discovered
-				_, cloudProviderFound := currentCloudInstances[node.Name]
-				if _, found := registeredNodes[nodeName]; found && !cloudProviderFound {
-					// Confirm that node is deleted by cloud provider, instead of
-					// a not-autoscaled node
-					nodeExists, existsErr := csr.cloudProvider.NodeExists(node)
-					if existsErr == nil && !nodeExists {
-						nodesRemoved = append(nodesRemoved, node)
-					}
-				}
-			}
-
-			// Seek nodes that may have been deleted since last update
-			// cloudProviderNodeInstances are retrieved by nodeGroup,
-			// not autoscaled nodes will be excluded
-			for _, instances := range csr.cloudProviderNodeInstances {
-				for _, instance := range instances {
-					if _, found := currentCloudInstances[instance.Id]; !found {
-						// Check Kubernetes registered nodes for corresponding deleted
-						// Cloud Provider instance
-						if kubeNode, kubeNodeFound := registeredNodes[instance.Id]; kubeNodeFound {
-							// Confirm that node is deleted by cloud provider, instead of
-							// a not-autoscaled node
-							nodeExists, existsErr := csr.cloudProvider.NodeExists(kubeNode)
-							if existsErr == nil && !nodeExists {
-								nodesRemoved = append(nodesRemoved, kubeNode)
-							}
-						}
-					}
-				}
-			}
+	for _, node := range allNodes {
+		if !csr.hasCloudProviderInstance(node) {
+			nodesRemoved = append(nodesRemoved, node)
 		}
 	}
 	return nodesRemoved
 }
 
+func (csr *ClusterStateRegistry) hasCloudProviderInstance(node *apiv1.Node) bool {
+	exists, err := csr.cloudProvider.HasInstance(node)
+	if err == nil {
+		return exists
+	}
+	if !errors.Is(err, cloudprovider.ErrNotImplemented) {
+		klog.Warningf("Failed to check whether node has cloud instance for %s: %v", node.Name, err)
+		return exists
```

Comment:
I don't think depending on the value of `exists` makes sense here. If the function returned an error, this is likely going to be some const. Maybe just fall through here and do the taint check instead?

Reply:
I think falling back to the taint check would be against expectations.  I believe if the cloud provider only implements the check for certain nodes, then falling back to taint check would make sense.  However, if there is an error that is something other than `ErrNotImplemented`, then I believe the expectation would be to return the default value of `false` and log the error.  If taint based detection is used as a fallback, then it could cause conflicting issues in the event that errors are transient.  I feel that the goal is to migrate the cloud providers to the new detection methodology, and the taint-based version is there for backwards compatibility.

Please let me know your thoughts.

Reply:
I think if cloudprovider isn't able to tell (for whatever reason: lack of implementation or some runtime error) whether the instance exists, clusterstate can fall back to the next best method of checking this, which is checking taints. In what scenario do you think it would be better to blindly use `false` instead of inspecting the taint?

Reply:
After thinking it over for a bit, I believe you're correct.  Initially I wanted to avoid confusing the check between the new method and old taint-based version.  However, referring back to an earlier point you raised about allowing cloud providers to implement the new check selectively by node/instance type/etc., it makes sense to fallback to the taint-based when the other option fails.  My previous understanding was that the taint-based check would be replaced entirely, but that might not work for all cloud providers.  I'll work on an update that includes falling back to the taint-based check.